utils/src/bitmap/mod.rs

```diff
@@ -0,0 +1,1705 @@
+            if self.get_bit(i) != other.get_bit(i) {
+                return false;
+            }
+        }
+
+        true
+    }
+}
+
+impl<const N: usize> Eq for BitMap<N> {}
+
+impl<T: AsRef<[bool]>, const N: usize> From<T> for BitMap<N> {
+    fn from(t: T) -> Self {
+        let bools = t.as_ref();
+        let mut bv = Self::new();
+        for &b in bools {
+            bv.push(b);
+        }
+        bv
+    }
+}
+
+impl<const N: usize> From<BitMap<N>> for Vec<bool> {
+    fn from(bv: BitMap<N>) -> Self {
+        bv.iter().collect()
+    }
+}
+
+impl<const N: usize> fmt::Debug for BitMap<N> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
+        // For very large BitMaps, only show a preview
+        const MAX_DISPLAY: usize = 64;
+        const HALF_DISPLAY: usize = MAX_DISPLAY / 2;
+
+        // Closure for writing a bit
+        let write_bit = |formatter: &mut Formatter<'_>, index: u64| -> core::fmt::Result {
+            formatter.write_char(if self.get_bit(index) { '1' } else { '0' })
+        };
+
+        f.write_str("BitMap[")?;
+        let len = self.len();
+        if len <= MAX_DISPLAY as u64 {
+            // Show all bits
+            for i in 0..len {
+                write_bit(f, i)?;
+            }
+        } else {
+            // Show first and last bits with ellipsis
+            for i in 0..HALF_DISPLAY as u64 {
+                write_bit(f, i)?;
+            }
+
+            f.write_str("...")?;
+
+            for i in (len - HALF_DISPLAY as u64)..len {
+                write_bit(f, i)?;
+            }
+        }
+        f.write_str("]")
+    }
+}
+
+impl<const N: usize> Index<usize> for BitMap<N> {
+    type Output = bool;
+
+    /// Allows accessing bits using the `[]` operator.
+    ///
+    /// Panics if out of bounds.
+    #[inline]
+    fn index(&self, index: usize) -> &Self::Output {
+        self.assert_index(index as u64);
+        let value = self.get_bit(index as u64);
+        if value {
+            &true
+        } else {
+            &false
+        }
+    }
+}
+
+impl<const N: usize> BitAnd for &BitMap<N> {
+    type Output = BitMap<N>;
+
+    fn bitand(self, rhs: Self) -> Self::Output {
+        self.assert_eq_len(rhs);
+        let mut result = self.clone();
+        result.and(rhs);
+        result
+    }
+}
+
+impl<const N: usize> BitOr for &BitMap<N> {
+    type Output = BitMap<N>;
+
+    fn bitor(self, rhs: Self) -> Self::Output {
+        self.assert_eq_len(rhs);
+        let mut result = self.clone();
+        result.or(rhs);
+        result
+    }
+}
+
+impl<const N: usize> BitXor for &BitMap<N> {
+    type Output = BitMap<N>;
+
+    fn bitxor(self, rhs: Self) -> Self::Output {
+        self.assert_eq_len(rhs);
+        let mut result = self.clone();
+        result.xor(rhs);
+        result
+    }
+}
+
+impl<const N: usize> Write for BitMap<N> {
+    fn write(&self, buf: &mut impl BufMut) {
+        // Prefix with the number of bits
+        self.len().write(buf);
```

Comment:
Using varint encoding here (and in `encode_size`) is likely more efficient. This would automatically be taken care of if `len()` returns `usize` instead of `u64`